lib/launcher.rb

```diff
@@ -62,6 +69,10 @@ def https_app(https_port, plugins = https_plugins)
         logger.error "Unable to read #{settings.ssl_ca_file}. Are the values correct in settings.yml and do permissions allow reading?"
       end
 
+      unless File.readable?(settings.foreman_ssl_ca)
```

Comment:
I just noticed you're suing `foreman_ssl_ca` but you should use `ssl_ca_file`. `foreman_ssl_ca` is used to communicate with Foreman, while falling back to `ssl_ca_file`. `ssl_ca_file` is used to serve. So this check would be redundant.

Reply:
Hello @ekohl 

I'm using `foreman_ssl_ca` intentionally.  `ssl_ca_file` will always contain only the internal CA. When a deployment is using self-signed certs, `ssl_ca_file` and `foreman_ssl_ca` are identical and contain a certificate chain that can validate `ssl_certificate`

~~~
# openssl verify -CAfile ./foreman_ssl_ca.pem ssl_cert.pem 
ssl_cert.pem: OK
# openssl verify -CAfile ./ssl_ca.pem ssl_cert.pem 
ssl_cert.pem: OK
~~~

However, when a deployment uses custom certificates, then `ssl_ca_file` only contains internal CA while `foreman_ssl_ca `contains the CA bundle which forms the full certificate chain for `ssl_certificate`

~~~
# openssl verify -CAfile ./foreman_ssl_ca.pem ssl_cert.pem 
ssl_cert.pem: OK
# openssl verify -CAfile ./ssl_ca.pem ssl_cert.pem 
CN = satellite.jpasqualetto.local
error 20 at 0 depth lookup: unable to get local issuer certificate
error ssl_cert.pem: verification failed
~~~

So, if we add `ssl_ca_file` into `SSLExtraChainCert`  this will work fine for deployments with self-signed certs but will break on deployments with custom certs.

If we add `foreman_ssl_ca` into `SSLExtraChainCert` it works for both use cases.

With all this being said, the check to see if we can read `foreman_ssl_ca `seems necessary, just as checking `ssl_ca_file` is.

Reply:
Sorry, but if that's the case then that's just a bug in how the installer deploys things then. The `foreman_ssl_ca` is the CA used to verify the connection to Foreman. `ssl_ca_file` is the CA that must match the public and private key. I'm very firm in saying that it must be `ssl_ca_file`.

Looking at the certs we have this bit that deploys it:
https://github.com/theforeman/puppet-certs/blob/5d7679a1add25a087d8dd925c1bae11d003852b5/manifests/foreman_proxy.pp#L95-L115

If `$proxy_key` and `$proxy_cert` are signed by `$server_ca_cert`, then `$proxy_ca_cert` should also use `$server_ca_cert` as a source.

Now that I look closer, we already send `SSLCACertificateFile` so perhaps this whole PR is not needed, but the real bug is in puppet-certs.

Reply:
I think https://github.com/theforeman/puppet-certs/pull/413 is the actual fix for this bug.

Reply:
@ehelms pointed out we also do both verification of client certs (which is the default CA), so it's one I need to think further about.

Reply:
Note to self: perhaps we can provide the cert including the full chain in `SSLCertificate`.